pkg/controller/service/servicecontroller.go

```diff
@@ -120,42 +164,44 @@ func New(cloud cloudprovider.Interface, kubeClient clientset.Interface, clusterN
 //
 // It's an error to call Run() more than once for a given ServiceController
 // object.
-func (s *ServiceController) Run(serviceSyncPeriod, nodeSyncPeriod time.Duration) error {
-	if err := s.init(); err != nil {
+func (sc *ServiceController) Run(workers int, stopCh <-chan struct{}) error {
+	if err := sc.init(); err != nil {
 		return err
 	}
 
-	// We have to make this check beecause the ListWatch that we use in
-	// WatchServices requires Client functions that aren't in the interface
-	// for some reason.
-	if _, ok := s.kubeClient.(*clientset.Clientset); !ok {
+	if _, ok := sc.kubeClient.(*clientset.Clientset); !ok {
 		return fmt.Errorf("ServiceController only works with real Client objects, but was passed something else satisfying the clientset.Interface.")
 	}
+	defer runtime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	nodeLW := cache.NewListWatchFromClient(sc.kubeClient.(*clientset.Clientset).CoreClient, "nodes", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(nodeLW, &api.Node{}, sc.nodeLister.Store, 0).Run()
+	go sc.nodeSyncLoop(nodeSyncPeriod)
```

Comment:
This part was not changed as there is a time.Tick look inside nodeSyncLoop function, so I left it as no changed, I will change it as you suggested anyway.